### PR TITLE
completions: add completions for bash and fish

### LINKED ISSUE #615: shell tab completions
I've been building a track, so I've been using configlet a lot. With the plethora of subcommands and options, I figured tab completion would be handy. My shell is fish, so that's where I started.

But then I thought "how to install the completions"? How about this: add an option to `fetch-configlet` to download a shell-specific completions file, and emit instructions on how to source it into the user's shell.

Example:
```console
$ bin/fetch-configlet -c fish
To enable completions: source bin/configlet-completions.fish
$ bin/fetch-configlet -c bash
To enable completions: . bin/configlet-completions.bash
$ ls -1 bin
configlet
configlet-completions.bash
configlet-completions.fish
fetch-configlet
...
```

Do you think this would be useful?

### PR DESCRIPTION
**Pull request by glennj**
Not being a zsh or PS user, those completions are not included.

Changes to fetch-configlet to detect user's shell and print instructions will follow in a separate PR.

Closes #615

### CODE REVIEW

**Comment by ErikSchierboom** [comment]
Thanks! I'll let ee7 review this, as he is much more well-versed into this than I am.

File: scripts/fetch-configlet [hunk 1]
```diff
@@ -49,6 +49,30 @@ get_download_url() {
     cut -d'"' -f4
 }
 
+# get the user's shell (as portably as possible)
+# and print out the command to enable tab completion
+emit_completion_command() {
+    local shell file
+    if command -v getent >/dev/null; then
+        shell=$(getent passwd "${USER}" | cut -d: -f6)
```
Comment (ee7) [inline scripts/fetch-configlet]:
```suggestion
        shell=$(getent passwd "${USER}" | cut -d: -f7)
```

The format for `/etc/passwd` is the below, right?
```
account:password:UID:GID:GECOS:directory:shell
```

So we need the seventh field to get the shell?

For example, on Arch Linux:

```console
$ getent passwd "${USER}"
myusername:x:1000:1000::/home/myusername:/bin/zsh
$ getent passwd "${USER}" | cut -d: -f6
/home/myusername
$ getent passwd "${USER}" | cut -d: -f7
/bin/zsh 
```

On OpenBSD:

```console
$ getent passwd "${USER}"
myusername:*:1000:1000:myusername:/home/myusername:/bin/ksh
```

File: scripts/fetch-configlet [hunk 2]
```diff
@@ -49,6 +49,30 @@ get_download_url() {
...
+    elif command -v finger >/dev/null; then
+        shell=$(finger "${USER}" | sed -nE '/.*Shell: / { s///; s/ .*//; p; }')
+    elif [[ -f /etc/passwd ]]; then
+        shell=$(awk -F: -v u="${USER}" '$1 == u {print $6}' /etc/passwd)
```
Comment (ee7) [inline scripts/fetch-configlet]:
```suggestion
    elif [[ -f /etc/passwd ]]; then
        shell=$(awk -F: -v u="${USER}" '$1 == u {print $7}' /etc/passwd)
```

Ditto.

File: completions/configlet.bash [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    return 1
+}
+
+_configlet_complete_lint_() {
+    _configlet_complete_options_ "$global_opts"
+}
+
+_configlet_complete_generate_() {
+    _configlet_complete_options_ "$global_opts"
+}
```
Comment (ee7) [inline completions/configlet.bash]: For later: `configlet generate` will (soon-ish) have the same non-destructive-by-default behavior as `configlet fmt`, and so support the same options (`-u`, `-y`, `-e`) . So we should remember to update the completions accordingly.

File: completions/configlet.bash [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+_configlet_complete_sync_() {
+    case $prev in
+        '--tests')
+            _configlet_complete_options_ "choose include exclude"
+            ;;
+        '-e' | '--exercise')
+            _configlet_complete_slugs_ "practice"
+            ;;
+        *)
+            _configlet_complete_options_ "-e --exercise -o --offline -u --update -y --yes --docs --metadata --tests $global_opts"
```
Comment (ee7) [inline completions/configlet.bash]:
```suggestion
            _configlet_complete_options_ "-e --exercise -o --offline -u --update -y --yes --docs --filepaths --metadata --tests $global_opts"
```

Was missing `--filepaths`?

File: completions/configlet.fish [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+complete -c configlet -n "__fish_seen_subcommand_from fmt" -s e -l exercise -d "exercise slug" \
+    -x -a "(find ./exercises/{concept,practice} -maxdepth 1 -mindepth 1 -type d -printf '%P\n' | sort)"
+
+# sync subcommand
+complete -c configlet -n "__fish_use_subcommand" -a sync -d "Check or update Practice Exercise docs, metadata, and tests"
+complete -c configlet -n "__fish_seen_subcommand_from sync" -s o -l offline -f -d "Do not update proc-spec cache"
+complete -c configlet -n "__fish_seen_subcommand_from sync" -s u -l update -d "Write changes"
+complete -c configlet -n "__fish_seen_subcommand_from sync" -s y -l yes -d "Auto-confirm update"
+complete -c configlet -n "__fish_seen_subcommand_from sync"      -l docs -d "Sync docs only"
+complete -c configlet -n "__fish_seen_subcommand_from sync"      -l metadata -d "Sync metadata only"
```
Comment (ee7) [inline completions/configlet.fish]:
```suggestion
complete -c configlet -n "__fish_seen_subcommand_from sync"      -l docs -d "Sync docs only"
complete -c configlet -n "__fish_seen_subcommand_from sync"      -l filepaths -d "Sync filepaths only"
complete -c configlet -n "__fish_seen_subcommand_from sync"      -l metadata -d "Sync metadata only"
```

Ditto.

**Review by ee7 — COMMENTED** [review]
Thanks for working on this. Very cool.

First pass at a review. Great that it's even got slug completion, but does it find the slugs at "tab time" or at source time? The latter would be a problem when working on multiple tracks. (I haven't tested yet, sorry).

~I see that [`db12f3b..10dd1e0`](https://github.com/exercism/configlet/compare/db12f3b5b38287e2c79f38c0bec261ded0d4358c..10dd1e032b22c0196dfcd609ca19c590ef8b7104) removed the shell detection - are you leaning towards not trying to detect it?~

Edit: Oh, from https://github.com/exercism/configlet/issues/615#issuecomment-1181410546 I guess you'll make a separate PR for the `fetch-configlet` changes.

**Comment by glennj** [comment]
> does it find the slugs at "tab time" or at source time? 

At tab time.

File: completions/configlet.bash (hunk 4 above)
Comment (glennj) [reply completions/configlet.bash]: Woopsie

File: completions/configlet.bash (hunk 3 above)
Comment (glennj) [reply completions/configlet.bash]: When that happens, we can just dispatch:
```bash
 _configlet_complete_generate_() {
    _configlet_complete_fmt_
}
```

File: scripts/fetch-configlet (hunk 1 above)
Comment (glennj) [reply scripts/fetch-configlet]: Good catch. Will fix in next PR.

**Review by ee7 — APPROVED** [review]
Looks great. Thanks again, and sorry again for the delay in reviewing.

I hadn't used fish for a long time. Looks like they've improved compatibility a lot lately (e.g. support for `&&` and `||`, `&>`, and `foo=bar echo $foo`). And the completion approach for fish seems nice. Clear and self-documenting.

I took the liberty of pushing a few trivial commits (please yell at me if you object), and I've opened a bunch of follow-up issues. Minor things like https://github.com/exercism/configlet/issues/637 (I don't know if it's easy to address) and #642.

Merging.

**Comment by glennj** [comment]
Thanks @ee7!

**Comment by ee7** [comment]
Released at last: https://github.com/exercism/configlet/releases/tag/4.0.0-beta.6

**Pull request merged**